### PR TITLE
Introduce "RunException" which invokes the stop_run api.

### PR DESCRIPTION
**Pull request by vdbergh**
Also: fix (hopefully) verify_signature().

Also: replace p.returncode by format_return_code(p.returncode)
everywhere.

Also: give feedback in the event log if a user is not authorized
to stop a run.

### CODE REVIEW

**Comment by vdbergh** [comment]
I changed this to WIP since I think it would be better to leave the formatting of the `stop_run` error message to server (as it is done for `failed_task`) but I am a bit busy now.

**Comment by vdbergh** [comment]
Removed WIP label.

**Comment by ppigazzini** [comment]
DEV and workers updated.

**Comment by vdbergh** [comment]
Did you really update DEV? `stop_run()` is now supposed to print a message in the event log even if the user is not authorized to actually stop the run. I just tested it locally but it does not seem to work on DEV.

**Comment by ppigazzini** [comment]
Yes, here is the log snippet:
```
From https://github.com/glinscott/fishtest
 * branch            refs/pull/1157/head -> FETCH_HEAD
Merge made by the 'recursive' strategy.
 server/fishtest/api.py     | 28 ++++++++++----
 server/tests/test_api.py   |  6 ++-
 server/tests/test_users.py |  2 +-
 worker/games.py            | 95 +++++++++++++++++++++++++++-------------------
 worker/worker.py           | 11 ++++--
 5 files changed, 90 insertions(+), 52 deletions(-)
Requirement already satisfied: pip in ./fishtest/server/env/lib/python3.8/site-p                                                                                                                               ackages (21.1.1)
```

**Comment by ppigazzini** [comment]
If I'm reading correctly the code you haven't added a new API requiring a nginx config update.

**Comment by vdbergh** [comment]
No. `api_stop_run` is now: 
```
    def stop_run(self):
        self.require_authentication()
        username = self.get_username()
        user = self.request.userdb.user_cache.find_one({"username": username})
        if not user:
            return {}
        if self.task_id() is None:
            return {}
        authorized = True
        if user["cpu_hours"] < 1000:
            authorized = False
        with self.request.rundb.active_run_lock(self.run_id()):
            run = self.request.rundb.get_run(self.run_id())
            message = self.request.json_body.get("message", "API request")
            run["stop_reason"] = "task_id: {}, worker: {}, reason: '{}' {}".format(
                self.task_id(),
		worker_name(run["tasks"][self.task_id()]["worker_info"]),
                message[:1024],
                " (not authorized)" if not authorized else "",
            )
            run_ = del_tasks(run)
            self.request.actiondb.stop_run(username, run_)
            if authorized:
                run["finished"] = True
                run["failed"] = True
                self.request.rundb.stop_run(self.run_id())
        return {}
```
I don't see what's wrong (and it works locally).

**Comment by vdbergh** [comment]
I did not introduce a new api but the option part of the url changes of course.

**Comment by vdbergh** [comment]
I wonder if there are any exceptions in the log...

**Comment by ppigazzini** [comment]
```
sudo journalctl -u fishtest@6543 --since "4 hours ago"
...
Nov 16 13:48:02 vps1572612706 systemd[1]: Stopped Fishtest Server port 6543.
Nov 16 13:48:43 vps1572612706 systemd[1]: Started Fishtest Server port 6543.
Nov 16 13:50:33 vps1572612706 pserve[25517]: Failed_task: failure for: https://tests.stockfishchess.org/tests/view/61892411d7a085ad008ef299, task_id: 1600, worker: user01-3cores-db6c288d, reason: 'Terminated
Nov 16 13:52:44 vps1572612706 pserve[25517]: Failed_task: failure for: https://tests.stockfishchess.org/tests/view/618a6914d7a085ad008ef3ad, task_id: 592, worker: user01-3cores-c26e1e5e, reason: 'Terminated
Nov 16 13:53:13 vps1572612706 pserve[25517]: Failed_task: failure for: https://tests.stockfishchess.org/tests/view/618927eed7a085ad008ef2a3, task_id: 1671, worker: user01-3cores-ef4cbfe0, reason: 'Terminated
Nov 16 13:55:21 vps1572612706 pserve[25517]: dead task: run: https://tests.stockfishchess.org/tests/view/61892411d7a085ad008ef299 task_id: 1601 worker: user01-3cores-c26e1e5e
Nov 16 13:57:20 vps1572612706 pserve[25517]: dead task: run: https://tests.stockfishchess.org/tests/view/618ad66ad7a085ad008ef3ed task_id: 351 worker: user01-3cores-b0d5bf30
Nov 16 14:35:31 vps1572612706 pserve[25517]: Failed_task: failure for: https://tests.stockfishchess.org/tests/view/618a6914d7a085ad008ef3ad, task_id: 593, worker: user01-3cores-cb7e7c2c, reason: 'Terminated
Nov 16 14:35:40 vps1572612706 systemd[1]: Stopping Fishtest Server port 6543...
Nov 16 14:35:40 vps1572612706 pserve[25517]: flush
Nov 16 14:35:40 vps1572612706 pserve[25517]: done
Nov 16 14:35:40 vps1572612706 pserve[25517]: Starting server in PID 25517.
Nov 16 14:35:41 vps1572612706 systemd[1]: Stopped Fishtest Server port 6543.
Nov 16 14:36:22 vps1572612706 systemd[1]: Started Fishtest Server port 6543.
Nov 16 14:39:27 vps1572612706 pserve[28620]: dead task: run: https://tests.stockfishchess.org/tests/view/6188b47cd7a085ad008ef23b task_id: 1651 worker: user01-3cores-cb7e7c2c
Nov 16 14:42:27 vps1572612706 pserve[28620]: Failed_task: failure for: https://tests.stockfishchess.org/tests/view/618ad66ad7a085ad008ef3ed, task_id: 352, worker: user01-3cores-4a0fca7f, reason: 'Terminated
Nov 16 14:42:27 vps1572612706 pserve[28620]: dead task: run: https://tests.stockfishchess.org/tests/view/618a6914d7a085ad008ef3ad task_id: 594 worker: user01-3cores-2f5f4070
Nov 16 14:42:47 vps1572612706 pserve[28620]: dead task: run: https://tests.stockfishchess.org/tests/view/6188b47cd7a085ad008ef23b task_id: 1652 worker: user01-3cores-2f5f4070
Nov 16 14:45:12 vps1572612706 pserve[28620]: dead task: run: https://tests.stockfishchess.org/tests/view/618927eed7a085ad008ef2a3 task_id: 1672 worker: user01-3cores-4a0fca7f
Nov 16 15:57:42 vps1572612706 systemd[1]: Stopping Fishtest Server port 6543...
Nov 16 15:57:42 vps1572612706 pserve[28620]: flush
Nov 16 15:57:42 vps1572612706 pserve[28620]: done
Nov 16 15:57:42 vps1572612706 pserve[28620]: Starting server in PID 28620.
Nov 16 15:57:43 vps1572612706 systemd[1]: Stopped Fishtest Server port 6543.
Nov 16 15:57:43 vps1572612706 systemd[1]: Started Fishtest Server port 6543.

```

**Comment by ppigazzini** [comment]
Tonight I will debug DEV.

**Comment by vdbergh** [comment]
Thanks. It should be a trivial thing.

**Comment by ppigazzini** [comment]
The problem on DEV seems to be due to this code:
```python
        user = self.request.userdb.user_cache.find_one({"username": username})
        if not user:
            print("api_stop_run: Invalid user {}".format(username))
```
`games.py` patched to trigger a wrong bench check  (`if int(bench_sig) != int(signature) + 1:`)

With a worker ran by a new user ("user01" created by script, or "user02" registered and unblocked) the server doesn't fin the username in the user_cache and returns immediately 
```
api_stop_run: Invalid user user02
```

With a my fishtest ran by my fishtest username the server stops the tasks.

This is weird because I always tested that api using "user01" or "user00" and skipping by hand the check on the min hours contributed.

**Comment by ppigazzini** [comment]
BTW a CRTL-C during the time sleep stops the worker but there is not an entry in the DEV events log:

```
Exception running games:
Wrong bench in stockfish_c58708732e2710c752f049c08f9e3b67be71bd66 Expected: 6652223 Got: 6652223
Informing the server
Task exited
Waiting 240.0 seconds before retrying
  Send heartbeat for 40993a78-c117-429a-ba6a-c3ed344444fb ... Pleased to hear from you...
  Send heartbeat for 40993a78-c117-429a-ba6a-c3ed344444fb ... Pleased to hear from you...
^CTraceback (most recent call last):
  File "/home/usr00/_git/__test_folder00/fishtest/worker/worker.py", line 726, in <module>
    sys.exit(worker())
  File "/home/usr00/_git/__test_folder00/fishtest/worker/worker.py", line 714, in worker
    time.sleep(delay)
  File "/home/usr00/_git/__test_folder00/fishtest/worker/worker.py", line 225, in on_sigint
    raise WorkerException("Terminated by signal {}".format(str_signal(signal)))
games.WorkerException: Terminated by signal SIGINT
Heartbeat stopped
Deleting lock file /home/usr00/_git/__test_folder00/fishtest/worker/worker.lock
```

**Comment by vdbergh** [comment]
> BTW a CRTL-C during the time sleep stops the worker but there is not an entry in the DEV events log:
> 
> ```
> Exception running games:
> Wrong bench in stockfish_c58708732e2710c752f049c08f9e3b67be71bd66 Expected: 6652223 Got: 6652223
> Informing the server
> Task exited
> Waiting 240.0 seconds before retrying
>   Send heartbeat for 40993a78-c117-429a-ba6a-c3ed344444fb ... Pleased to hear from you...
>   Send heartbeat for 40993a78-c117-429a-ba6a-c3ed344444fb ... Pleased to hear from you...
> ^CTraceback (most recent call last):
>   File "/home/usr00/_git/__test_folder00/fishtest/worker/worker.py", line 726, in <module>
>     sys.exit(worker())
>   File "/home/usr00/_git/__test_folder00/fishtest/worker/worker.py", line 714, in worker
>     time.sleep(delay)
>   File "/home/usr00/_git/__test_folder00/fishtest/worker/worker.py", line 225, in on_sigint
>     raise WorkerException("Terminated by signal {}".format(str_signal(signal)))
> games.WorkerException: Terminated by signal SIGINT
> Heartbeat stopped
> Deleting lock file /home/usr00/_git/__test_folder00/fishtest/worker/worker.lock
> ```

Ctrl-C is only logged when it occurs in `run_games()`.  BTW in the very latest commit, I replaced time.sleep() by a wrapper `safe_sleep()` at some places to get a nicer output when the use presses Ctrl-C.

**Comment by ppigazzini** [comment]
`delta_update_users.py` was disabled on DEV (leftover of #902 wip), I'm trying to check if this changes something.

**Comment by vdbergh** [comment]
So `user00` and `user01` are in the `userdb` (since authentication works) but not in the cache.

The cache check seems pointless though since the user is already authenticated. api_failed_task does not do a cache check (of course it would be nice to understand why the user is not in the cache).

**Comment by vdbergh** [comment]
> So `user00` and `user01` are in the `userdb` (since authentication works) but not in the cache.
> 
> The cache check seems pointless though since the user is already authenticated. api_failed_task does not do a cache check (of course it would be nice to understand why the user is not in the cache).

Ah but we need properties of the user.

**Comment by vdbergh** [comment]
I just checked that user00 and user01 are in the user cache on localhost.

```
$ mongo
...
> use fishtest_new
switched to db fishtest_new
> db.user_cache.find().pretty()
{
	"_id" : ObjectId("5fdf2b6ab8517fd296720f7d"),
	"username" : "user01",
	"cpu_hours" : 20.918919111111123,
	"games" : 24640,
	"tests" : 43,
	"tests_repo" : "https://github.com/official-stockfish/Stockfish",
	"last_updated" : "1 days ago",
	"games_per_hour" : 0,
	"task_last_updated" : ISODate("2021-11-16T10:18:04.455Z"),
	"diff" : 96177.810408
}
{
	"_id" : ObjectId("607b1edabd58b4aa44b69046"),
	"username" : "user00",
	"cpu_hours" : 0,
	"games" : 0,
	"tests" : 2,
	"tests_repo" : "https://github.com/official-stockfish/Stockfish",
	"last_updated" : "Never",
	"games_per_hour" : 0,
	"diff" : 86400000000000
}
```

**Comment by vdbergh** [comment]
`delta_update_users.py` seems to be the only place where there is actually something inserted in `user_cache`.

**Comment by ppigazzini** [comment]
40 minutes ago I loaded a fresh DB, ran by hand some `delta_update_users.py` and one `delta_update_users.py all`, but I continue to have not the new users in the cache.
```
user=None
api_stop_run: Invalid user user01
user=None
api_stop_run: Invalid user user00
```

I should investigate also these exceptions....
```
${VENV}/bin/python3 delta_update_users.py
Exception on:
'task_last_updated'
Exception on:
'task_last_updated'
Exception on:
'task_last_updated'
Exception on:
'task_last_updated'
Exception on:
'task_last_updated'
Exception on:
'task_last_updated'
Exception on:
'task_last_updated'
Exception on:
'task_last_updated'
Exception on:
'task_last_updated'
...
```

**Comment by vdbergh** [comment]
I wonder what is actually the purpose of "user_cache"...  It seems to be an extract of "users".

**Comment by ppigazzini** [comment]
My account works fine:
```
user=None
api_stop_run: Invalid user user01
user={'_id': ObjectId('125e08eef51002efd7a3a98c6'), 'username': 'ctoks', 'cpu_hours': 378465.6008060935, 'games': 25147877, 'tests': 3, 'tests_repo': '', 'last_updated': '2 hours ago', 'games_per_hour': 0.0, 'task_last_updated': datetime.datetime(2021, 11, 17, 11, 5, 28, 868000), 'diff': 8323.883033}
Update_task: task 8 is not active
dead task: run: https://tests.stockfishchess.org/tests/view/6194f8a3a5e44985837e31f0 task_id: 8 worker: user00-3cores-d77184f1
user=None
api_stop_run: Invalid user user01
user=None
api_stop_run: Invalid user user01
```

**Comment by ppigazzini** [comment]
Now suddenly it started working...
```
user=None
api_stop_run: Invalid user user01
dead task: run: https://tests.stockfishchess.org/tests/view/6194f89ca5e44985837e31ee task_id: 9 worker: user01-3cores-6eb7b2b0
dead task: run: https://tests.stockfishchess.org/tests/view/6194f8b0a5e44985837e31f4 task_id: 9 worker: user01-3cores-6eb7b2b0
dead task: run: https://tests.stockfishchess.org/tests/view/6194f8a3a5e44985837e31f0 task_id: 10 worker: user01-3cores-3447c93e
dead task: run: https://tests.stockfishchess.org/tests/view/6194f8b5a5e44985837e31f6 task_id: 9 worker: user01-3cores-3447c93e
user={'_id': ObjectId('61950427991dab4de114d7c7'), 'username': 'user01', 'cpu_hours': 0.6869333333333335, 'games': 800, 'tests': 1, 'tests_repo': 'https://github.com/official-stockfish/Stockfish', 'last_updated': '3 minutes ago', 'games_per_hour': 6378.979643827638, 'task_last_updated': datetime.datetime(2021, 11, 17, 13, 28, 5, 390000), 'diff': 194.291687}
api_stop_run: User user01 has too few games to stop a run
```

**Comment by vdbergh** [comment]
Yes now it also works for me...

**Comment by ppigazzini** [comment]
user00 is still missing after joining some workers with this username...
```
user=None
api_stop_run: Invalid user user00
user=None
api_stop_run: Invalid user user00
```
- Users
 
Username | Last active | Games/Hour | CPU Hours | Games played | Tests submitted | Tests repository
-- | -- | -- | -- | -- | -- | --
user01 | 1 hours ago | 6377 | 0 | 800 | 1 | https://github.com/official-stockfish/Stockfish
vdbergh | 2 hours ago | 0 | 137062 | 8928293 | 122 | https://github.com/official-stockfish/Stockfish

- Users - Top Month

Username | Last active | Games/Hour | CPU Hours | Games played | Tests submitted | Tests repository
-- | -- | -- | -- | -- | -- | --
user00 | seconds ago | 13934 | 4 | 4936 | 0 |  
user01 | seconds ago | 6377 | 13 | 15144 | 5 | https://github.com/official-stockfish/Stockfish

**Comment by vdbergh** [comment]
Did you check the collection `user_cache` (instead of `users`)?
 
For me `delta_update_users.py` runs without errors (there are some easily fixable deprecation warnings).

However I do not really understand by what mechanism delta_update_users.py is run.

**Comment by ppigazzini** [comment]
> Did you check the collection `user_cache` (instead of `users`)?
Yes, reloaded the DB half hour ago

```
> db.user_cache.find({username: "ctoks"}).pretty()
{
        "_id" : ObjectId("61756f583c0d736e5b3e7cae"),
        "username" : "ctoks",
        "cpu_hours" : 378010.5761660914,
        "games" : 25136725,
        "tests" : 3,
        "tests_repo" : "",
        "last_updated" : "10 hours ago",
        "games_per_hour" : 0,
        "task_last_updated" : ISODate("2021-11-17T06:34:04.334Z"),
        "diff" : 36733.987116
}
> db.user_cache.find({username: "user00"}).pretty()
> db.user_cache.find({username: "user01"}).pretty()
```

> For me `delta_update_users.py` runs without errors (there are some easily fixable deprecation warnings).

I tried to fix the deprecation warnings, but I'm an absolute beginner with pymongo, please review #902 :)

> However I do not really understand by what mechanism delta_update_users.py is run.

```
fishtest@tests:~$ crontab -l
VENV=/home/fishtest/fishtest/server/env
UPATH=/home/fishtest/fishtest/server/utils

# Backup mongodb database every 6 hours and upload to s3
3 */6 * * * /usr/bin/cpulimit -l 50 -f -m -- sh ${UPATH}/backup.sh

# Update the users table
1,16,31,46 * * * * ${VENV}/bin/python3 ${UPATH}/delta_update_users.py

# Purge old pgn files
33 3 * * * ${VENV}/bin/python3 ${UPATH}/purge_pgn.py

# Clean up old mail (more than 9 days old)
32 5 * * * screen -D -m mutt -e 'push D~d>9d<enter>qy<enter>'
```

**Comment by ppigazzini** [comment]
DEV and workers updated.

**Comment by ppigazzini** [comment]
DEV and workers updated.

**Comment by ppigazzini** [comment]
DEV and workers updated.

File: worker/games.py [hunk 1]
```diff
@@ -387,7 +400,9 @@ def make_targets():
 
     if p.returncode != 0:
         raise WorkerException(
-            "make help failed with return code {}".format(p.returncode)
+            "make help failed with return code {}".format(
+                format_return_core(p.returncode)
```
Comment (ppigazzini) [inline worker/games.py:404]: Typo? "_core"

**Comment by ppigazzini** [comment]
https://docs.python.org/3/library/subprocess.html#popen-objects
Reference for the future, to get the return code within the context manager block: 
- `Popen.poll()` loop:
```python
import subprocess
with subprocess.Popen(["ls", "-la"]) as p:
    while p.poll() is None:
        pass
    print(f"inner: {p.returncode=}")
    print(f"{p.poll()=}")
print(f"outer: {p.returncode=}")
print(f"{p.poll()=}")
```
-  `Popen.wait()`:
```
import subprocess
with subprocess.Popen(["ls", "-la"]) as p:
    p.wait()
    print(f"inner: {p.returncode=}")
print(f"outer: {p.returncode=}")
```
-  `Popen.communicate()`:
```python
import subprocess
with subprocess.Popen(["ls", "-la"]) as p:
    p.communicate()
    print(f"inner: {p.returncode=}")
print(f"outer: {p.returncode=}")
```

File: worker/games.py (hunk 1 above)
Comment (vdbergh) [reply worker/games.py:404]: Yep...

**Comment by ppigazzini** [comment]
DEV and workers updated.

**Comment by ppigazzini** [comment]
DEV and workers updated.

**Comment by ppigazzini** [comment]
Merged with cli after reformatting with latest black & isort.
PROD updated, triggered the workers update.
Thank you @vdbergh :)

**Pull request closed without merging**